DanmakuEngine/Movements/LinearMovement.cs

```diff
@@ -0,0 +1,79 @@
+using System.Numerics;
+using DanmakuEngine.Bindables;
+using Silk.NET.Maths;
+
+namespace DanmakuEngine.Movements;
+
+public abstract class LinearMovement<T> : AccumulatedMovement<T>, ICanGetSpeed<T>, ICanSetSpeed<T>
+    where T : IEquatable<T>
+{
+    protected double currentPeriodStartTime;
+
+    private T _speed = default!;
+
+    public T Speed
+    {
+        get => _speed;
+        set
+        {
+            if (_speed.Equals(value))
+                return;
+
+            _speed = value;
+
+            Active.BindValueChanged(splitPeriod, true);
+
+            void splitPeriod(ValueChangedEvent<bool> e)
+            {
+                if (e.NewValue)
+                {
+                    currentPeriodStartTime = CurrentTime;
+
+                    Active.ValueChanged -= splitPeriod;
+                }
+            }
+        }
+    }
+
+    public LinearMovement(T speed)
+    {
+        this.Speed = speed;
+    }
+}
+
+public class LinearMovementF(float speed)
+    : LinearMovement<float>(speed)
+{
+    protected override float AccumulatedValue()
+        => Value.Value +
+           // we still want to try to avoid accumulating to prevent floating point errors
+           Speed * (float)(CurrentTime - currentPeriodStartTime);
```

Comment:
any suggestion?

Reply:
Can you apply `Kahan summation algorithm` to both `Value.Value` and `Speed` calculations? 
And my previous may be misleading as i didn't use `startValue` by mistake. Generally it is used to split time into different periods so that we can set different speed when moving.
Here is the corrected code:
```Csharp
public abstract class LinearMovementBase<T> : AccumulatedMovement<T>, ICanGetSpeed<T>, ICanSetSpeed<T>
    where T : IEquatable<T>
{
    protected double currentPeriodStartTime;

    protected double currentPeriodElapsedTime
        => CurrentTime - currentPeriodStartTime;

    private T _speed = default!;

    public T Speed
    {
        get => _speed;
        set
        {
            if (_speed.Equals(value))
                return;

            _speed = value;

            currentPeriodStartTime = CurrentTime;
            startValue = Value.Value;
        }
    }

    protected abstract T compensation { get; set; }

    public LinearMovementBase(T speed)
    {
        this.Speed = speed;

        Active.BindValueChanged(e =>
        {
            if (e.NewValue)
                currentPeriodStartTime = CurrentTime;
        });
    }
}

public class LinearMovementF(float speed)
    : LinearMovementBase<float>(speed)
{
    protected override float compensation { get; set; } = 0;

    protected override float AccumulatedValue()
    {
        // implement this pls
    }
}
```

Reply:
You may misunderstood my code. We should not accumulate `startValue`. we made `Value.Value` to `AccumulatedValue()` in `Update()` when updating a frame. The `startValue` only records the initial value of each period and we calculate `AccumulatedValue` with it and how much we actually ran in current period(which is `Speed * currentPeriodElapsedTime`). Since we did't accumulate any value and we calculate the correct value every frame, I don't think `floating point errors` matter and don't need Kahan summation algorithm. What's your opinions?

Reply:
my corrected code, implemented as i just said.
```Csharp
    protected override double AccumulatedValue()
        => startValue +
           // we still want to try to avoid accumulating to prevent floating point errors
           Speed * (CurrentTime - currentPeriodStartTime);
```